### PR TITLE
Fontconfig support

### LINKED ISSUE #15: Configurable font
Awesome work with adding configuration file!

In the light of the new and exciting configurable wldash, could the font face also be configurable please?

### PR DESCRIPTION
**Pull request by LiHRaM**
Closes #15.

This draft experiments with using [`yeslogic-fontconfig`](https://lib.rs/crates/yeslogic-fontconfig) to load fonts.
The current design uses a blend of Mono and Sans typefaces to render the GUI.
I think this is a good approach, and would therefore like to suggest how we can keep that feature as we implement the configuration side.

I propose that
1. We add two options to the configuration file, one which defaults to a Sans typeface and another which defaults to a Mono typeface.
2. We configure widgets to default to the one or the other.

I'm not sure if we want to add the option of customizing typeface per widget, or if we want to give the less flexible, but easier to implement option of choosing between Sans/Mono.

Also, I'm not sure it has to be Sans/Mono, it could also be Primary/Secondary, since any font installed could be used for either.

## To Do:
- [x] Add runtime font loading
- [x] Add config support for font loading
- [x] Give useful error message when font is not supported by `rusttype`
- [x] Handle cases where users accidentally remove "mono" and "sans" from config?

### CODE REVIEW

**Comment by kennylevinsen** [comment]
This PR made me realize that fontconfig is making *awful* decisions for default mono/sans fonts on my machine. :|

I'm not sure I understand the reason behind fixing two fonts in place if we're making it configurable anyway. While I do agree that I had *impeccable* sense of design with for the hardcoded fonts, I assume that if people want to *mess up the masterpiece* anyway, we might as well give them the flexibility to set whatever.

Rather than fix it in place, we could for example do something like this: (psuedo-config)

```yaml
fonts:
  sans: DejaVu
  mono: Inconsolata
  clockface: WebDings
widgets:
  verticalLayout:
  - clock:  # Override font for clock, resulting in WebDings
      font: clockface
  - date: {} # Using default 'sans' font setting, which takes DejaVu
  - launcher: # Use mono font for launcher, which takes Inconsolata
      font: mono
```

Widgets default to font "sans", "mono". The `fonts` map can be used to override fonts, and each widget can have a `font` option to set what they should use (some widgets might have multiple font options, e.g. calendar).

From a configuration perspective, it looks the same as your suggestion if you only want two fonts, but with no limitations for how many font one uses. It should also be pretty easy to implement.

(Also, I noted that some of the widgets' size calculations seem slightly borked when you change the font metrics significantly, so we'll probably get a few follow-up issues from making things configurable.)

**Comment by LiHRaM** [comment]
@kennylevinsen a few thoughts: The configuration style that you suggested is really nice, good idea!

I wasn't quite able to figure out how to get it working straight away without boxing everything, but it works now. :smile: 

Also, rusttype seems unhappy with many of my `.otf` fonts, even though their project README says they support them. Any idea what's happening there?

For example, it will complain about `SF Mono Powerline` and `Operator Mono` being "Ill Formed"

```
wldash on  font_support is 📦 v0.2.0 via 🦀 v1.44.0 
λ cargo run
   Compiling wldash v0.2.0 (/home/lihram/git/github.com/kennylevinsen/wldash)
    Finished dev [unoptimized + debuginfo] target(s) in 3.55s
     Running `/home/lihram/.cargo/target_dir/debug/wldash`
thread 'main' panicked at 'Loading /home/lihram/.local/share/fonts/SF-Mono-Powerline-Regular.otf failed: IllFormed', src/main.rs:87:13
```

File: src/config.rs [hunk 1]
```diff
@@ -68,49 +77,76 @@ impl Widget {
         self,
         time: NaiveDateTime,
         tx: Sender<Cmd>,
+        fonts: FontMap,
```
Comment (kennylevinsen) [inline src/config.rs]: This should be a reference, as it doesn't need to mutate, and to avoid cloning.

File: src/config.rs [hunk 2]
```diff
@@ -68,49 +77,76 @@ impl Widget {
...
     ) -> Option<Box<dyn widget::Widget + Send>> {
+        let get_font = |font_name: &str| match fonts.get(font_name) {
+            Some(f) => Box::new(f.clone()),
```
Comment (kennylevinsen) [inline src/config.rs]: This is not great. It causes every font-user to clone the font into their own box, and the FontMap itself is also cloned a lot. For X fonts with Y users and Z vert/horiz layouts (due to cloned maps), we end up with X+XZ+Y font instances in Y boxes. With my config, I believe that adds up to 18 font copies even though I'm only using 2 fonts! And that's before counting all the extra clones occuring in some of the widget constructors.

It should be possible to use references to a single font, but might require manual lifetime annotation. Might be able to cheat with a 'static bound.

(Also a little weird to have this be a closure here.)

File: src/config.rs [hunk 3]
```diff
@@ -68,49 +77,76 @@ impl Widget {
...
-            } => match widget.construct(time, tx) {
+            } => match widget.construct(time, tx, fonts) {
                 Some(w) => Some(widget::Fixed::new((width, height), w)),
                 None => None,
             },
             Widget::HorizontalLayout(widgets) => Some(widget::HorizontalLayout::new(
                 widgets
                     .into_iter()
-                    .map(|x| x.construct(time, tx.clone()))
+                    .map(|x| x.construct(time, tx.clone(), fonts.clone()))
```
Comment (kennylevinsen) [inline src/config.rs]: Should be able to get rid of this with a FontMap reference.

File: src/config.rs [hunk 4]
```diff
@@ -68,49 +77,76 @@ impl Widget {
...
                     .filter(|x| x.is_some())
                     .map(|x| x.unwrap())
                     .collect(),
             )),
             Widget::VerticalLayout(widgets) => Some(widget::VerticalLayout::new(
                 widgets
                     .into_iter()
-                    .map(|x| x.construct(time, tx.clone()))
+                    .map(|x| x.construct(time, tx.clone(), fonts.clone()))
```
Comment (kennylevinsen) [inline src/config.rs]: Same

File: src/widgets/date.rs [hunk 5]
```diff
@@ -11,8 +14,8 @@ pub struct Date {
 }
 
 impl Date {
-    pub fn new(time: NaiveDateTime, size: f32) -> Box<Date> {
-        let mut date_cache = Font::new(&ROBOTO_REGULAR, size);
+    pub fn new(time: NaiveDateTime, font: FontRef, size: f32) -> Box<Date> {
+        let mut date_cache = Font::new(font.clone(), size);
```
Comment (kennylevinsen) [inline src/widgets/date.rs]: Unnecessary clone.

File: src/draw.rs [hunk 6]
```diff
@@ -99,7 +85,7 @@ impl Font {
     pub fn add_str_to_cache(&mut self, s: &str) {
         for ch in s.chars() {
             if self.glyphs.get(&ch).is_none() {
-                let glyph = CachedGlyph::new(self.font, self.size, ch);
+                let glyph = CachedGlyph::new(self.font.clone(), self.size, ch);
```
Comment (kennylevinsen) [inline src/draw.rs]: Unwanted clone - this one leads to a new font copy every time we prime a font cache.

File: src/widgets/calendar.rs [hunk 7]
```diff
@@ -153,14 +156,20 @@ impl Calendar {
...
+        time: NaiveDateTime,
+        font_primary: FontRef,
+        font_secondary: FontRef,
+        font_size: f32,
+        sections: u32,
+    ) -> Box<Calendar> {
+        let mut calendar_cache = Font::new(font_secondary, font_size * 2.0);
         calendar_cache.add_str_to_cache("0123456789");
-        let mut month_cache = Font::new(&ROBOTO_REGULAR, font_size * 4.0);
+        let mut month_cache = Font::new(font_primary.clone(), font_size * 4.0);
```
Comment (kennylevinsen) [inline src/widgets/calendar.rs]: Unwanted clone. Should be fixed if we make the fonts references again.

File: src/widgets/calendar.rs [hunk 8]
```diff
@@ -153,14 +156,20 @@ impl Calendar {
...
         month_cache.add_str_to_cache("JanuryFebMchApilJgstSmOoNvD");
-        let mut year_cache = Font::new(&DEJAVUSANS_MONO, font_size * 1.5);
+        let mut year_cache = Font::new(font_primary.clone(), font_size * 1.5);
```
Comment (kennylevinsen) [inline src/widgets/calendar.rs]: Same.

**Review by kennylevinsen — CHANGES_REQUESTED** [review]
> @kennylevinsen a few thoughts: The configuration style that you suggested is really nice, good idea!
> 
> I wasn't quite able to figure out how to get it working straight away without boxing everything, but it works now. smile

It's always good to have a starting point. :)

The main problem isn't the boxing, though, but the cloning. We end up with a *huge* amount of font copies, so we need to find a fix for that. We should be able to make give the FontMap a lifetime long enough to be able to use references everywhere, though.

> Also, rusttype seems unhappy with many of my `.otf` fonts, even though their project README says they support them. Any idea what's happening there?
> 
> For example, it will complain about `SF Mono Powerline` and `Operator Mono` being "Ill Formed"

Have you tried bumping rusttype to 0.9? I think some work was done towards better .otf support.

**Comment by LiHRaM** [comment]
@kennylevinsen I've refactored it so that there is no requirement for a 'static lifetime, and am now using references. However, it's complaining that the lifetime is not long enough. Maybe you have some ideas for how to make this work?

**Comment by kennylevinsen** [comment]
Good work - I don't mind taking a look, as long as it can wait a little. My plate is a bit full as the moment. :)

**Comment by LiHRaM** [comment]
Just found this, and think it might be relevant: https://gitlab.redox-os.org/redox-os/rusttype/issues/142.
Looks like cloning Fonts is not so expensive after all?

**Comment by kennylevinsen** [comment]
Ugh. That's unexpected, and a little ugly. I feel like reference counting should be left entirely up to the library *user* by giving them only owned types, but it appears that rusttype disagreed.

Looking at the source, Font is just an enum covering an Arc over the font, so cloning (without boxing) is fine. Sorry for the extra work. :/

**Comment by LiHRaM** [comment]
Yeah, I agree, it is a little ugly. But I'm sure they have their reasons.

Anyway, I was able to get it working using just references. There's a couple of drawbacks though:
- I brought in crossbeam, which didn't seem to bring in any extra transitive dependencies
- There's a lot of explicit lifetime annotations now

I'm happy making that tradeoff though, considering that we avoid cloning the font hashmap.

**Comment by LiHRaM** [comment]
Okay, it's ready for review now. I'd forgotten to remove the fonts. :smile:

cc @kennylevinsen

**Review by kennylevinsen — APPROVED** [review]

**Comment by kennylevinsen** [comment]
Squashed, slightly adjusted, and merged.

Notes: For some reason, fontconfig pulls in a bunch of dependencies, including things like freetype. This can be seen with `ldd`. Also, for some reason, using the same fonts now through fontconfig seem to result in different metrics - the fonts are smaller when using the same size. :/

**Pull request closed without merging**